sdks/python/apache_beam/utils/windowed_value.py

```diff
@@ -279,6 +293,208 @@ def create(value, timestamp_micros, windows, pane_info=PANE_INFO_UNKNOWN):
+    return HomogeneousWindowedBatch(
+        WindowedValue(values, timestamp, windows, pane_info))
+
+  @property
+  def values(self):
+    return self._wv.value
+
+  @property
+  def timestamp(self):
+    return self._wv.timestamp
+
+  @property
+  def pane_info(self):
+    return self._wv.pane_info
+
+  @property
+  def windows(self):
+    return self._wv.windows
+
+  @windows.setter
+  def windows(self, value):
+    self._wv.windows = value
+
+  def with_values(self, new_values):
+    # type: (Any) -> WindowedBatch
+    return HomogeneousWindowedBatch(self._wv.with_value(new_values))
+
+  def as_windowed_values(self, explode_fn: Callable) -> Iterable[WindowedValue]:
+    for value in explode_fn(self._wv.value):
+      yield self._wv.with_value(value)
+
+  def __eq__(self, other):
+    if isinstance(other, HomogeneousWindowedBatch):
+      return self._wv == other._wv
+    return NotImplemented
+
+  def __hash__(self):
+    return hash(self._wv)
+
+
+class ConcreteWindowedBatch(WindowedBatch):
+  """A concrete WindowedBatch where all event-time information is stored
+  independently for each element.
+
+  Attributes:
+    values: The underlying values of the windowed batch.
+    timestamp: An iterable of timestamps associated with the value as seconds
+      since Unix epoch.
+    windows: An iterable with a set (iterable) of window objects for each value.
+      The window objects are descendants of the BoundedWindow class.
+    pane_info: An iterable of PaneInfo descriptors describing the triggering
+      information for the pane that contained each value. Alternatively, a
+      single PaneInfo may be specified to use for every value. If None, will be
+      set to PANE_INFO_UNKNOWN.
+  """
+  def __init__(
+      self,
+      values,
+      timestamps,  # type: Sequence[TimestampTypes]
+      windows,  # type: Iterable[Tuple[BoundedWindow, ...]]
+      pane_infos=PANE_INFO_UNKNOWN  # type: Union[Iterable[PaneInfo],PaneInfo]
+  ):
+    self.values = values
+
+    def convert_timestamp(timestamp: TimestampTypes) -> int:
+      if isinstance(timestamp, int):
+        return timestamp * 1000000
+      else:
+        # TODO: Cache Timestamp object as in WindowedValue?
+        timestamp_object = (
+            timestamp
+            if isinstance(timestamp, Timestamp) else Timestamp.of(timestamp))
+        return timestamp_object.micros
+
+    self.timestamp_objects: Optional[List[Timestamp]] = None
+    self.timestamps_micros = [convert_timestamp(t) for t in timestamps]
+    self.windows = windows
+    #TODO: Should we store length?
+    #self.length = length
+    self.pane_infos = pane_infos
+
+  @property
+  def timestamps(self) -> Sequence[Timestamp]:
+    if self.timestamp_objects is None:
+      self.timestamp_objects = [
+          Timestamp(0, micros) for micros in self.timestamps_micros
+      ]
+
+    return self.timestamp_objects
+
+  def with_values(self, new_values):
```

Comment:
Presumably there's some constraints on the size and ordering of new_values?

Reply:
Actually, does this ever get used?

Reply:
No this implementation is never used, since everything is built on HomogeneousWindowedBatch for now.

Reply:
Removed along with all of ConcreteWindowedBatch